### PR TITLE
Provisioning SPI changes

### LINKED ISSUE #15: Separate binding from queue provisioning
Downstream changes from https://github.com/spring-cloud/spring-cloud-stream/issues/745

### PR DESCRIPTION
**Pull request by sobychacko**
This PR depends on https://github.com/spring-cloud/spring-cloud-stream/pull/787
 
- Make the necessary changes in binder to adjust with the provisioning SPI introduced in core
 - Fixing tests
 - logging in provisioner
 - cleanup

Fixes #15

### CODE REVIEW

File: spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java [hunk 1]
```diff
@@ -65,26 +176,70 @@ private Queue createQueue(String topicName, Session session, String consumerName
...
-				JmsUtils.commitIfNecessary(session);
-				JmsUtils.closeSession(session);
-				JmsUtils.closeConnection(connection);
-			} catch (JMSException e) {
-				e.printStackTrace();
+				return partitionTopics.get(-1).getTopicName();
+			}
+			catch (JMSException e) {
+				if (logger.isInfoEnabled()) {
+					logger.info("JMS Exception", e);
```
Comment (artembilan) [inline spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java]: Why exception into an `info`?
There is no reason to check `logger.isInfoEnabled()`, if you don't use string concatenation in the log message:
```
public
  void log(String callerFQCN, Priority level, Object message, Throwable t) {
    if(repository.isDisabled(level.level)) {
      return;
    }
```
OTOH there is no exception any more.
Just `Map.get()`

File: spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java [hunk 2]
```diff
@@ -65,26 +176,70 @@ private Queue createQueue(String topicName, Session session, String consumerName
...
+		public String getPartitionedProducerDestinationName(int partition) {
+			try {
+				return partitionTopics.get(partition).getTopicName();
+			}
+			catch (JMSException e) {
+				if (logger.isInfoEnabled()) {
+					logger.info("JMS Exception", e);
+				}
 			}
+			return null;
```
Comment (artembilan) [inline spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java]: Same here: `map.get()`.
No `JMSException`, but can be `NPE` if there is no entry for the provided `partition`.
What is our action then?
Is it really appropriate to return `null` to the caller?
Maybe some `IllegalArgumentException` would be better ?

File: spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java [hunk 3]
```diff
@@ -44,18 +149,21 @@ public Destinations provisionTopicAndConsumerGroup(String topicName, String... c
...
-		} catch (JMSException e) {
-			e.printStackTrace();
+			if (groups != null) {
+				return groups[0];
+			}
 		}
-
-		return new Destinations(topic, groups);
+		catch (JMSException e) {
+			throw new IllegalStateException(e);
```
Comment (garyrussell) [inline spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java]: Spring has an unchecked `JmsException`.

File: spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java [hunk 4]
```diff
@@ -65,26 +173,61 @@ private Queue createQueue(String topicName, Session session, String consumerName
...
-				JmsUtils.commitIfNecessary(session);
-				JmsUtils.closeSession(session);
-				JmsUtils.closeConnection(connection);
-			} catch (JMSException e) {
-				e.printStackTrace();
+				return partitionTopics.get(-1).getTopicName();
 			}
+			catch (JMSException e) {
+				throw new IllegalStateException(e);
+			}
```
Comment (garyrussell) [inline spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java]: ditto - other places too.

File: spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/config/ActiveMQJmsConfiguration.java [hunk 5]
```diff
@@ -53,8 +56,9 @@ public ActiveMQConnectionFactory connectionFactory(ActiveMQConfigurationProperti
 	}
 
 	@Bean
-	public ActiveMQQueueProvisioner solaceQueueProvisioner(ActiveMQConnectionFactory connectionFactory) throws Exception {
-		return new ActiveMQQueueProvisioner(connectionFactory);
+	ProvisioningProvider solaceQueueProvisioner(ActiveMQConnectionFactory connectionFactory,
```
Comment (garyrussell) [inline spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/config/ActiveMQJmsConfiguration.java]: Not your issue, but why are we naming beans about Solace in ActiveMQ ?

File: spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java (hunk 4 above)
Comment (artembilan) [inline spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java]: M-m-m. There was my comment that we don't have here any JMS exception at all: https://github.com/spring-cloud/spring-cloud-stream-binder-jms/pull/16#discussion_r99956485

File: spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java [hunk 6]
```diff
@@ -8,34 +27,120 @@
...
+			connection = connectionFactory.createConnection();
+			session = connection.createSession(true, 1);
+			session.createQueue(ACTIVE_MQ_DLQ);
+		}
+		catch (JMSException e) {
+			throw new IllegalStateException(e);
+		}
+		finally {
+			try {
+				JmsUtils.commitIfNecessary(session);
```
Comment (garyrussell) [inline spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java]: This should be in its own try/catch so the closes will be reliably called; even if we get a `JMSException` on the commit.

File: spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java (hunk 4 above)
Comment (garyrussell) [inline spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java]: FYI, the compiler won't let you catch a checked exception unless some statement in the try block can throw one - "Unreachable catch block...".

File: spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java (hunk 3 above)
Comment (garyrussell) [inline spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java]: I though we agreed to use a generic `ProvisioningException` here - did I miss a decision change?

Comment (sobychacko) [inline spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java]: No, you did not. My bad, corrected and pushed a new change.

File: spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/config/ActiveMQJmsConfiguration.java (hunk 5 above)
Comment (garyrussell) [inline spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/config/ActiveMQJmsConfiguration.java]: Looks like this wasn't addressed - also, `ProvisioningProvider` needs `<?, ?>`

File: spring-cloud-stream-binder-jms-activemq/src/test/java/org/springframework/cloud/stream/binder/jms/activemq/RepublishMessageRecovererTests.java [hunk 7]
```diff
@@ -63,27 +56,29 @@ public static void initTests() throws Exception {
...
-		target.recover(createMessage(ImmutableMap.of("fancy", "header")), cause);
-		message = jmsTemplate.receive(DEAD_LETTER_QUEUE);
+//	@Test
+//	public void recover_provisionsDeadLetterQueue() throws Exception {
+//		target.recover(createMessage(ImmutableMap.of("fancy", "header")),DEAD_LETTER_QUEUE, cause);
+//		message = jmsTemplate.receive(DEAD_LETTER_QUEUE);
+//
+//		verify(consumerDestination, times(1)).getDlq();
+//	}
+
```
Comment (garyrussell) [inline spring-cloud-stream-binder-jms-activemq/src/test/java/org/springframework/cloud/stream/binder/jms/activemq/RepublishMessageRecovererTests.java]: Why is this test (and the Before) commented out?

**Pull request closed without merging**